File: Src/ExpectationLib/OrderedExpectedObservationsPattern.cs
namespace ExpectationLib;

/// <summary>
/// A pattern matcher that expects observations to meet a list expectation that must occur in order
/// <remarks>Other observations that do not meet expectation can occur in between</remarks>
/// </summary>
public class OrderedExpectedObservationsPattern : IExpectatedObservationsPattern
{
    /// <summary>
    ///  A matcher that expects observations to meet a list expectation that must occur in order
    /// </summary>
    /// <param name="expectations">Expectations where each must not match before the prior expectation</param>
    /// <param name="observations">observations that must match the list of ordered expectations</param>
    public OrderedExpectedObservationsPattern(List<IExpectation> expectations, List<Observation> observations)
    {
        Expectations = expectations;
        Observations = observations;
    }

    /// <summary>
    /// Expectations
    /// </summary>
    public List<Observation> Observations { get;  }

    /// <summary>
    /// Expectations
    /// </summary>
    public List<IExpectation> Expectations { get; }

    /// <summary>
    /// Unmatched expectations
    /// </summary>
    public List<IExpectation> UnmatchedExpectations => Expectations.Except(MatchedExpectations).ToList();

    /// <summary>
    /// Matched expectations
    /// </summary>
    public List<IExpectation> MatchedExpectations { get; } = new();

    /// <inheritdoc />
    public bool Match()
    {
        var countMatches = 0;
        int i = 0;
        
        foreach (var observation in Observations)
        {
            if (i > Expectations.Count-1 ) continue;
            if (!Expectations[i].Match(observation)) continue; // can have other unmatched observations in between
            MatchedExpectations.Add(Expectations[i]);                
            i++;
            countMatches++;
        }

        return countMatches == Expectations.Count;
    }
}

File: Src/CppLibExpectation/OrderedExpectedObservationsPattern.cpp
#include "OrderedExpectedObservationsPattern.h"

namespace ExpectationLib
{
	OrderedExpectedObservationsPattern::OrderedExpectedObservationsPattern(
		const std::vector<std::shared_ptr<IExpectation>>& expectations,
		const std::vector<std::shared_ptr<Observation>> observations)
	{
		Expectations = expectations;
		Observations = observations;
	}

	std::vector<std::shared_ptr<IExpectation>> OrderedExpectedObservationsPattern::UnmatchedExpectations()
	{
		std::vector<std::shared_ptr<IExpectation>> unmatched;
		for (auto expectation : Expectations) 
		{
			if (std::find(MatchedExpectations.begin(), MatchedExpectations.end(), expectation) == MatchedExpectations.end()) 
			{
				unmatched.push_back(expectation);
			}
		}
		return unmatched;
	}

	bool OrderedExpectedObservationsPattern::Match()
	{
		int countMatches = 0;
		int i = 0;
		for (auto observation : Observations) 
		{
			if (i > Expectations.size() - 1) continue;
			if (!Expectations[i]->Match(observation)) continue; // can have other unmatched observations in between
			MatchedExpectations.push_back(Expectations[i]);
			i++;
			countMatches++;
		}
		return countMatches == Expectations.size();
	}
}
